lib/sensu/server/tessen.rb

```diff
@@ -0,0 +1,165 @@
+require "em-http-request"
+require "sensu/constants"
+
+module Sensu
+  module Server
+    class Tessen
+      attr_accessor :settings, :logger, :redis, :options
+      attr_reader :timers
+
+      # Create a new instance of Tessen. The instance variable
+      # `@timers` is use to track EventMachine timers for
+      # stopping/shutdown.
+      #
+      # @param options [Hash] containing the Sensu server Settings,
+      #   Logger, and Redis connection.
+      def initialize(options={})
+        @timers = []
+        @settings = options[:settings]
+        @logger = options[:logger]
+        @redis = options[:redis]
+        @options = @settings.to_hash.fetch(:tessen, {})
+      end
+
+      # Determine if Tessen is enabled (opt-out).
+      #
+      # @return [TrueClass, FalseClass]
+      def enabled?
+        @options[:enabled] != false
+      end
+
+      # Run Tessen, scheduling data reports (every 6h).
+      def run
+        schedule_data_reports
+      end
+
+      # Stop Tessen, cancelling and clearing timers.
+      def stop
+        @timers.each do |timer|
+          timer.cancel
+        end
+        @timers.clear
+      end
+
+      # Schedule data reports, sending data to the Tessen service
+      # immediately and then every 6 hours after that.
+      def schedule_data_reports
+        note = "this data helps inform the sensu team about installations"
+        note << " - you can choose to opt-out via configuration: {\"tessen\": {\"enabled\": false}}"
+        @logger.info("sending anonymized data to the tessen call-home service", :note => note)
+        send_data
+        @timers << EM::PeriodicTimer.new(21600) do
+          send_data
+        end
+      end
+
+      # Send data to the Tessen service.
+      def send_data(&block)
+        create_data do |data|
+          tessen_api_request(data, &block)
+        end
+      end
+
+      # Create data to be sent to the Tessen service.
+      #
+      # @return [Hash]
+      def create_data
+        get_install_id do |install_id|
+          get_client_count do |client_count|
+            get_server_count do |server_count|
+              token = @options.fetch(:token, "")
+              type, version = get_version_info
+              timestamp = Time.now.to_i
+              data = {
+                :token => token,
+                :install => {
+                  :id => install_id,
+                  :type => type,
+                  :version => version
```

Comment:
I assume this in only the server version, it also might be a good idea to include a hash of number of clients by client version.